### ISSUE TITLE
Any plans for Canonical Peer Priority

### ISSUE THREAD

**Issue by norzak** [webtorrent/webtorrent#1406]
Any plans to implement new bittorrent enhancement titled "Canonical Peer Priority"? which connects to geographically closer peers?

**Comment by SilentBot1** [comment]
Hey @norzak,

Just to make it easier for others, the BEP referenced is [BEP40](http://www.bittorrent.org/beps/bep_0040.html). 

This seems like it could possibly be implemented within the `._drain` function of the TCPOutgoingPeer queue `_queue` within [torrent.js](https://github.com/webtorrent/webtorrent/blob/925bb881cfe90cc055f67e747df693f32d265733/lib/torrent.js#L1672), prioritising the draining order based on priority calculated via the formula outlined in BEP40.

We would also need a way to determine the clients ip address as I don't believe we currently have a way. 

I would love for somebody to double check this to make sure I'm not missing something.

**Comment by norzak** [comment]
Hi @SilentBot1,

On the webtorrent.io page, during streaming video from the peers, I see some IP addresses connecting to mine in a visual animation. Where are those IPs coming from?

**Comment by feross** [comment]
This feature seems interesting, but not high priority. We're already one of the fastest torrent clients out there. I'm skeptical that we need to prioritize based on geographic location when the torrent protocol already prioritizes peers which upload the fastest.

The DDOS protection features seem interesting, though, so I've re-opened this issue.

**Comment by KayleePop** [comment]
I don't believe that BEP40 has anything to do with geographic location. Here's a good article about why it exists https://blog.libtorrent.org/2012/12/swarm-connectivity/

For WebRTC peers, I think this would have to be implemented in the tracker to have any benefit (because I don't think clients can choose which WebRTC peers to connect to, they just ask the tracker for one).

**Comment by feross** [comment]
@KayleePop Thanks for the link. I agree we should implement this! At least for TCP/uTP connections. Eventually, we can see about redesigning the WebTorrent tracker protocol to include a way for peers to get IP info before deciding to connect.
